src/crypto/pgp.rs

```diff
@@ -0,0 +1,418 @@
+use crate::crypto;
+use crate::errors::RPMError;
+
+use super::*;
+
+use std::io::Cursor;
+
+use ::pgp::{
+    composed::Deserializable,
+    types::KeyTrait,
+};
+
+fn now() -> ::chrono::DateTime<::chrono::Utc> {
+    // accuracy of serialized format is only down to seconds
+    use ::chrono::offset::TimeZone;
+    let now = ::chrono::offset::Utc::now();
+    ::chrono::offset::Utc.timestamp(now.timestamp(), 0u32)
+}
+
+/// Signer implementation using the `pgp` crate.
+///
+/// Note that this only supports ascii armored key files
+/// commonly with the file extension `.asc` as generated
+/// by i.e. `gpg`.
+#[derive(Clone, Debug)]
+pub struct Signer {
+    secret_key: ::pgp::composed::signed_key::SignedSecretKey,
+}
+
+impl crypto::Signing<crypto::algorithm::RSA> for Signer {
+    type Signature = Vec<u8>;
+    fn sign(&self, data: &[u8]) -> Result<Self::Signature, RPMError> {
+        let passwd_fn = || String::new();
+
+
+        let now = now();
+        let mut sig_cfg_bldr = ::pgp::packet::SignatureConfigBuilder::default();
+        let sig_cfg = sig_cfg_bldr
+            .version(::pgp::packet::SignatureVersion::V4)
+            .typ(::pgp::packet::SignatureType::Binary)
+            .pub_alg(::pgp::crypto::public_key::PublicKeyAlgorithm::RSA)
+            .hash_alg(::pgp::crypto::hash::HashAlgorithm::SHA2_256)
+            .issuer(Some(self.secret_key.key_id()))
+            .created(Some(now))
+            .unhashed_subpackets(vec![]) // must be initialized
+            .hashed_subpackets(vec![
+                ::pgp::packet::Subpacket::SignatureCreationTime(now),
+                ::pgp::packet::Subpacket::Issuer(self.secret_key.key_id()),
+                //::pgp::packet::Subpacket::SignersUserID("rpm"), TODO this would be a nice addition
+            ]) // must be initialized
+            .build()?;
+
+        let signature_packet = sig_cfg
+            .sign(&self.secret_key, passwd_fn, data)
+            .map_err(|e| format!("eee: {:?}", e))?;
+
+        let mut signature_bytes = Vec::with_capacity(1024);
+        let mut cursor = Cursor::new(&mut signature_bytes);
+        ::pgp::packet::write_packet(&mut cursor, &dbg!(signature_packet))
+            .map_err(|e| format!("eee: {:?}", e))?;
+
+        echo_signature("sign(pgp)", signature_bytes.as_slice());
```

Comment:
I think we should remove this too for the sake of noise

Reply:
you can leave the `echo_signature` function definition though because it might be helpful for debugging if one has to trace bugs.